### ISSUE TITLE
Recording rule with histogram_quantile fills in no-existent data

### ISSUE THREAD

**Issue by xujiaxj** [VictoriaMetrics/VictoriaMetrics#1058]
**Describe the bug**

For a sparse metric that only has data once in a while, the `histogram_quantile` function returns data with gaps as expected, but if I create a recording rule with this function, the recording rule returns continuous data. This also seems to mess up alerting rules's `for: 5m` clause randomly, i.e., we would expect an alert not fire if there is only one minute q99 data that violates a threshold, but no data in next four minutes. Instead, VM fires an alert.

**To Reproduce**
Create a Springboot application with a timer task that runs once every hour. You can use micrometer's `@Timed(histogram = true)` annotation to report metrics on this method. For example
```
public class MyTimerTask {

    // run once an hour by default
    @Scheduled(fixedDelayString = "${my.fixedDelay:3600000}",
               initialDelayString = "${my.initialDelay:30000}")
    @Timed(histogram = true)
    public void run() {
        // do something
    }
```

Now create a recording rule for its q99
```
- record: latency:q99
  expr: >
    histogram_quantile (
      0.99,
      sum(rate(method_timed_seconds_bucket[1m])>0) by (le, namespace, job, class, method, instance)
    )
```
Now run the following two queries in Grafana
```
2 * latency:q99{class="MyTimerTask", method="run"}
```
```
histogram_quantile (
            0.99,
            sum(rate(method_timed_seconds_bucket{class="MyTimerTask", method="run"}[1m])>0) by (le, namespace, job, class, method, instance)
          )
```

**Expected behavior**
Both time series will look similar like discrete dots spaced an hour apart, with the first one always double the value of the second one.

**Actual behavior**
Instead, we are seeing the first one drawing some continuous lines, while the second one having all the distinct dots. (Different colors are due to different instances.)

![image](https://user-images.githubusercontent.com/4541662/107077020-1f73a400-67a1-11eb-8bdb-694dfc37439d.png)

I checked the Chrome dev tool for the response data, and it indeed showed a continuous time series for the recording rule, so it's not a Grafana rendering issue. 

Another strange thing is that if I reduce the time range to <= 3 hours, the recording rule returns the discrete data points, much the same as its definition query returns, which is correct. 

**Version**
1.53.1

**Comment by xujiaxj** [comment]
Not sure if this is related to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/136

**Comment by hagen1778** [comment]
Hi @xujiaxj! This two approaches aren't compatible. Recording rule results in a new time series as result of query execution. It produces a new data point once every `evaluation_interval`, so if you have evaluation_interval=30s - it will result in a new data point every 30s. In contrary, the original time series `method_timed_seconds_bucket ` which you're using in the rule equation contains data points only once per hour (or what I can see from the screenshot). That's why, recording rule result is continuous and original query is not.

The second thing is that recording rules are executed against [`/api/v1/query`](https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries) endpoint, while grafana sends queries to [`/api/v1/query_range`](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries) endpoint. Where the former returns result based on the last recorded data point in series, and the latter returns result based on data points on the requested time range.

So to sum up, recording rule executes the given query with configured interval and stores every result of it in the form of a new time series. It will add a new data point once every evaluation_interval. You can reproduce this behavior by doing `curl` in a for loop against /api/v1/query endpoint with `sleep=evaluation_interval` per iteration.

**Comment by xujiaxj** [comment]
@hagen1778 Thanks for the reply. I am still a bit confused though.

The recording rule I have is essentially evaluating the same expression. If `method_timed_seconds_bucket` doesn't have a data point for a particular scrape interval, shouldn't the recording rule `latency:q99` also evaluates to no data point? Does it actually use `method_timed_seconds_bucket`'s data points older than the current evaluation interval?

On a separate note, is this related to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/383#issuecomment-605288117? By changing the `maxStalessInterval`, we were able to make the chart of the recording rule match exactly with the chart of the expression.

**Comment by hagen1778** [comment]
> By changing the maxStalessInterval, we were able to make the chart of the recording rule match exactly with the chart of the expression.

It is correct. VM assumes data points for every time series to appear in more or less fixed time intervals (no bigger than 5m usually). And if data point is missing, VM assumes that something "wrong" happened this time but the new data point will still arrive on the next interval. So it will try to auto-fill the gap with value of the previous recorded value to make Grafana panels looking good. It will try to do this up to some limit - usually the double of median time range between previous data points for this series. If no new data points were received and limit was reached - time series is deemed as "stale" and VM stops auto filling missing points. This is what you see for your original query.
When recording rule evaluates it follows the same behaviour. The only difference is that rule evaluation will produce more data points: every time it is evaluated and limit is not reached yet the new data point will be generated. You can verify it by exporting raw values for your metric and recording rule - the former will contain less data points than the latter. And because of this, recording rule result will look more continuous. 

To sum up I'd say that it is a bit of misuse for VM. It was designed to deal with time series, not events processing. Because of that, corner cases like this are happening. I'd also recommend to join our [Slack](http://slack.victoriametrics.com/) and check the following thread https://victoriametrics.slack.com/archives/CGZF1H6L9/p1611021094031400

**Comment by xujiaxj** [comment]
Just to be clear, when recording rules are evaluated, are auto-filled valued written to the storage? If at query time, there are gaps, does vmselect also auto-fill gaps in the query result?

**Comment by hagen1778** [comment]
`/api/v1/query` does not exactly "fill gaps", because there are no "range" to contain the gaps.
But `/api/v1/query` returns the last recorded data point (or its value processed by some function) for the given `time` param (current time is used if `time` not specified). And staleness here will define "how many seconds should pass without new points to consider time series as stale". Consider the following example:
```
time    10  20  30  40  50  60(now)
value   1   2   3   -   -   -

GET /api/v1/query?query=my_metric // maxStalessInterval auto-detected as avg range between points * 2  = 20s
returns nothing, since there are no data points in nearest 20s

GET /api/v1/query?query=my_metric // maxStalessInterval=60s set manually
returns 3, since it is the closest data point in 60s range
```

So for you metric, which gets a new data point about every ~30min, recording rule will produce a new data point during the next ~60min and only then will start to return empty results. Until the next data point will pop up.
Does it make sense?

**Comment by xujiaxj** [comment]
The difference between `/query` (recording rule) and `/query_range` can totally explain the difference seen in our Grafana chart. I think I get that. Thanks.

We did try the export query. However, it produces a single value for the recording rule, which is why I am still confused.
```
curl -G --data-urlencode 'match[]={__name__="latency:q99",class="MyTimerClass", method="run"}' 'https://chief.tsdb.dev.asserts.ai/select/0/prometheus/api/v1/export?&start=1612954770&end=1612955325' | jq -r .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   373  100   373    0     0   1695      0 --:--:-- --:--:-- --:--:--  1695
{
  "metric": {
    "__name__": "latency:q99",
    "namespace": "asserts",
    "job": "model-builder",
    "instance": "10.8.88.196:8060",
    "class": "MyTimerTask"
    "method": "run"
  },
  "values": [
    0.62008590204
  ],
  "timestamps": [
    1612954893075
  ]
}
```
Is this expected?

**Comment by hagen1778** [comment]
Hm, I feel confused as well) Could you pls attach the following things:
1. Screenshot of Grafana's /query_range for query `latency:q99{class="MyTimerClass", method="run"}` in at least 2h interval where you observe the issue;
2. Screenshot of Grafana's /query for query `histogram_quantile (0.99, sum(rate(method_timed_seconds_bucket{class="MyTimerClass", method="run"}[1m])>0) by (le, namespace, job, class, method, instance))`
3. Export of series for `latency:q99{class="MyTimerClass", method="run"}` and series for `method_timed_seconds_bucket{class="MyTimerClass", method="run"}` on the same time range.

Alternatively, you can export series (original and produced by rules) without sensitive data via [native export](https://github.com/VictoriaMetrics/VictoriaMetrics#how-to-import-data-in-native-format) and I'll try check it locally.

**Comment by xujiaxj** [comment]
Our environment has moved on. Since we have solved the problem by tuning `maxStalessInterval`, I think we can close this for now. If it comes up again, I can circle back. Thanks for the help!
